setup.py

```diff
@@ -1,25 +1,96 @@
-from distutils.core import setup, Extension
-import os.path, sys
-
-ale_c_lib = 'ale_python_interface/libale_c.so'
-if not os.path.isfile(ale_c_lib):
-  print('ERROR: Unable to find required library: %s. Please ensure you\'ve '\
-    'built ALE using CMake.'%(ale_c_lib))
-  sys.exit()
-
-module1 = Extension('ale_python_interface.ale_c_wrapper',
-                    libraries = ['ale_c'],
-                    include_dirs = ['src'],
-                    library_dirs = ['ale_python_interface'],
-                    extra_compile_args=['-D__STDC_CONSTANT_MACROS', '-std=c++11'],
-                    sources=['ale_python_interface/ale_c_wrapper.cpp'])
-setup(name = 'ale_python_interface',
-      version='0.6',
-      description = 'Arcade Learning Environment Python Interface',
-      url='https://github.com/bbitmaster/ale_python_interface',
-      author='Ben Goodrich',
-      license='GPL',
-      ext_modules = [module1],
-      packages=['ale_python_interface'],
-      package_dir={'ale_python_interface': 'ale_python_interface'},
-      package_data={'ale_python_interface': ['libale_c.so']})
+from setuptools import setup, Extension
+from setuptools.command.build_ext import build_ext
+import subprocess
+import platform
+import os
+
+
+class CMakeExtension(Extension):
+    def __init__(self, name, sourcedir="", config=[]):
+        Extension.__init__(self, name, sources=[])
+        self.sourcedir = os.path.abspath(sourcedir)
+        self.config = config
+
+
+class CMakeBuild(build_ext):
+    def build_extensions(self):
+        try:
+            subprocess.check_output(["cmake", "--version"])
+        except OSError:
+            raise RuntimeError(
+                "CMake must be installed to build the extensions: %s"
+                % ", ".join(ext.name for ext in self.extensions)
+            )
+
+        for ext in self.extensions:
+            extdir = os.path.abspath(os.path.dirname(self.get_ext_fullpath(ext.name)))
+            cfg = "Debug" if self.debug else "Release"
+
+            cmake_config_args = [
+                "-DOUTPUT_NAME={}".format(ext.name),
+                "-DCMAKE_BUILD_TYPE={}".format(cfg),
+                "-DCMAKE_LIBRARY_OUTPUT_DIRECTORY_{}={}".format(cfg.upper(), extdir),
+                "-DCMAKE_RUNTIME_OUTPUT_DIRECTORY_{}={}".format(cfg.upper(), extdir),
+                "-DCMAKE_ARCHIVE_OUTPUT_DIRECTORY_{}={}".format(
+                    cfg.upper(), self.build_temp
+                ),
+            ] + ext.config
+
+            cmake_build_args = []
+
+            # -DCMAKE_BUILD_TYPE doesn't work on Windows
+            # we need to specify --config Release at build type
+            if platform.system() == "Windows":
+                cmake_build_args += ["--config", cfg]
+
+            if not os.path.exists(self.build_temp):
+                os.makedirs(self.build_temp)
+
+            subprocess.check_call(
+                ["cmake", ext.sourcedir] + cmake_config_args, cwd=self.build_temp
```

Comment:
You should make sure you can do `pip install`, `pip install -e` and `python setup.py (install|develop)`, I remember having some issues where like only some of those would work so we had some crazy hack here: https://github.com/openai/atari-py/blob/master/setup.py#L17

Reply:
Such as https://github.com/openai/atari-py/issues/44 though I don't remember what that particular line fixes.

Reply:
All the different combinations of `pip install (-e) [ . | Arcade-Learning-Environment ]` work fine. I don't think `python Arcade-Learning-Environment/setup.py (develop|install)` should work? It won't be able to find the package data?

I did find one issue though. When doing develop builds with the editable flag it seems to expect the built library to have a platform-specific suffix. This seems to be new in Python 3, I'm not sure why this wasn't affecting `atari-py`. I have to resort to getting the proper suffix from `distutils.sysconfig` when building and at runtime.

Reply:
Yeah, I think `setup.py` has to be called when it's in the CWD.